tunix/sft/peft_trainer.py

```diff
@@ -312,12 +313,29 @@ def _shard_optimizer(self, mesh: shd.Mesh) -> None:
     """
     if mesh.empty:
       return
-    optimizer_state = nnx.state(
+    optimizer_state_arrays = nnx.state(
         self.optimizer, nnx.optimizer.OptState
     )  # select only the optimizer state
-    optimizer_shardings = nnx.get_named_sharding(optimizer_state, mesh)
+    optimizer_pspecs = nnx.get_partition_spec(optimizer_state_arrays)
+
+    def _adjust_pspec(pspec, state):
+      if pspec is None:
+        return None
+      # state can be a scalar, which is not an array.
+      state_ndim = getattr(state, 'ndim', 0)
+      if len(pspec) > state_ndim:
+        return shd.PartitionSpec(*pspec[:state_ndim])
+      return pspec
+
+    adjusted_pspecs = jax.tree.map(
+        _adjust_pspec, optimizer_pspecs, optimizer_state_arrays
+    )
+    optimizer_shardings = nn.logical_to_mesh_sharding(
+        adjusted_pspecs,
+        mesh,
+    )
```

Comment:
not sure I understand, what problem are you trying to solve here?

Reply:
I saw some of the state being scalar (for e.g., learning rate in the optimizer state) which is not a jax array. 
If the pspec is "too long" for the state, it's truncated to match the state's rank. For a scalar (state_ndim=0), this results in an empty PartitionSpec, P(), which correctly signifies that the scalar should be replicated across all devices.